File: src/main/java/com/github/redfoos/logstash/LogstashParserDefinition.java
package com.github.redfoos.logstash;

import com.github.redfoos.logstash.psi.LogstashFile;
import com.github.redfoos.logstash.psi.LogstashTypes;
import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.jetbrains.annotations.NotNull;

public class LogstashParserDefinition implements ParserDefinition {
    private static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    private static final TokenSet COMMENTS = TokenSet.create(LogstashTypes.COMMENT);
    private static final IFileElementType FILE = new IFileElementType(LogstashLanguage.getINSTANCE());

    @NotNull
    @Override
    public Lexer createLexer(Project project) {
        return new LogstashLexerAdapter();
    }

    @NotNull
    public TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @NotNull
    public TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @NotNull
    public TokenSet getStringLiteralElements() {
        return TokenSet.EMPTY;
    }

    @NotNull
    public PsiParser createParser(final Project project) {
        return new LogstashParser();
    }

    @Override
    public IFileElementType getFileNodeType() {
        return FILE;
    }

    public PsiFile createFile(FileViewProvider viewProvider) {
        return new LogstashFile(viewProvider);
    }

    @NotNull
    public PsiElement createElement(ASTNode node) {
        return LogstashTypes.Factory.createElement(node);
    }
}

File: src/main/kotlin/com/github/redfoos/logstash/LogstashParserDefinition.kt
package com.github.redfoos.logstash

import com.github.redfoos.logstash.psi.LogstashFile
import com.github.redfoos.logstash.psi.LogstashTypes
import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet

class LogstashParserDefinition: ParserDefinition {
    companion object {
        val COMMENTS = TokenSet.create(LogstashTypes.COMMENT)
        val FILE = IFileElementType(LogstashLanguage.INSTANCE)
        val STRING_LITERALS = TokenSet.create(LogstashTypes.SINGLE_QUOTED_STRING, LogstashTypes.DOUBLE_QUOTED_STRING)
    }

    override fun createLexer(project: Project?): Lexer = LogstashLexerAdapter()

    override fun createParser(project: Project?): PsiParser = LogstashParser()

    override fun getFileNodeType(): IFileElementType = FILE

    override fun getCommentTokens(): TokenSet = COMMENTS

    override fun getStringLiteralElements(): TokenSet = STRING_LITERALS

    override fun createElement(node: ASTNode?): PsiElement = LogstashTypes.Factory.createElement(node)

    override fun createFile(viewProvider: FileViewProvider): PsiFile = LogstashFile(viewProvider)

}
